File: src/main/java/org/mtransit/parser/gtfs/data/GCalendar.java
package org.mtransit.parser.gtfs.data;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Collection;
import java.util.List;

import org.mtransit.parser.Constants;

// https://developers.google.com/transit/gtfs/reference#calendar_fields
public class GCalendar {

	public static final String FILENAME = "calendar.txt";

	public static final String SERVICE_ID = "service_id";
	private String service_id;

	public static final String MONDAY = "monday";
	private boolean monday;

	public static final String TUESDAY = "tuesday";
	private boolean tuesday;

	public static final String WEDNESDAY = "wednesday";
	private boolean wednesday;

	public static final String THURSDAY = "thursday";
	private boolean thursday;

	public static final String FRIDAY = "friday";
	private boolean friday;

	public static final String SATURDAY = "saturday";
	private boolean saturday;

	public static final String SUNDAY = "sunday";
	private boolean sunday;

	public static final String START_DATE = "start_date";
	private int start_date; // YYYYMMDD

	public static final String END_DATE = "end_date";
	private int end_date; // YYYYMMDD

	private ArrayList<GCalendarDate> allDates;

	public GCalendar(String service_id, boolean monday, boolean tuesday, boolean wednesday, boolean thursday, boolean friday, boolean saturday, boolean sunday,
			int start_date, int end_date) {
		this.service_id = service_id;
		this.monday = monday;
		this.tuesday = tuesday;
		this.wednesday = wednesday;
		this.thursday = thursday;
		this.friday = friday;
		this.saturday = saturday;
		this.sunday = sunday;
		this.start_date = start_date;
		this.end_date = end_date;
		initAllDates();
	}

	public String getServiceId() {
		return service_id;
	}

	public boolean isServiceId(String serviceId) {
		return this.service_id.equals(serviceId);
	}

	public boolean isServiceIds(Collection<String> serviceIds) {
		return serviceIds.contains(this.service_id);
	}

	public int getStartDate() {
		return start_date;
	}

	public int getEndDate() {
		return end_date;
	}

	@Override
	public String toString() {
		return new StringBuilder() //
				.append(Constants.STRING_DELIMITER).append(this.service_id).append(Constants.STRING_DELIMITER).append(Constants.COLUMN_SEPARATOR) //
				.append(Constants.STRING_DELIMITER).append(this.monday).append(Constants.STRING_DELIMITER).append(Constants.COLUMN_SEPARATOR) //
				.append(Constants.STRING_DELIMITER).append(this.tuesday).append(Constants.STRING_DELIMITER).append(Constants.COLUMN_SEPARATOR) //
				.append(Constants.STRING_DELIMITER).append(this.wednesday).append(Constants.STRING_DELIMITER).append(Constants.COLUMN_SEPARATOR) //
				.append(Constants.STRING_DELIMITER).append(this.thursday).append(Constants.STRING_DELIMITER).append(Constants.COLUMN_SEPARATOR) //
				.append(Constants.STRING_DELIMITER).append(this.friday).append(Constants.STRING_DELIMITER).append(Constants.COLUMN_SEPARATOR) //
				.append(Constants.STRING_DELIMITER).append(this.saturday).append(Constants.STRING_DELIMITER).append(Constants.COLUMN_SEPARATOR) //
				.append(Constants.STRING_DELIMITER).append(this.sunday).append(Constants.STRING_DELIMITER).append(Constants.COLUMN_SEPARATOR) //
				.append(Constants.STRING_DELIMITER).append(this.start_date).append(Constants.STRING_DELIMITER).append(Constants.COLUMN_SEPARATOR) //
				.append(Constants.STRING_DELIMITER).append(this.end_date).append(Constants.STRING_DELIMITER).append(Constants.COLUMN_SEPARATOR) //
				.toString();
	}

	public boolean startsBefore(Integer date) {
		return this.start_date < date;
	}

	public boolean startsBetween(Integer startDate, Integer endDate) {
		return startDate <= this.start_date && this.start_date <= endDate;
	}

	public boolean isOverlapping(Integer startDate, Integer endDate) {
		return startsBetween(startDate, endDate) || endsBetween(startDate, endDate);
	}

	public boolean isInside(Integer startDate, Integer endDate) {
		return startsBetween(startDate, endDate) && endsBetween(startDate, endDate);
	}

	public boolean endsBetween(Integer startDate, Integer endDate) {
		return startDate <= this.end_date && this.end_date <= endDate;
	}

	public boolean endsAfter(Integer date) {
		return this.end_date > date;
	}

	public boolean containsDate(Integer date) {
		return this.start_date <= date && date <= this.end_date;
	}

	public List<GCalendarDate> getDates() {
		if (this.allDates == null) {
			initAllDates();
		}
		return this.allDates;
	}

	private void initAllDates() {
		this.allDates = new ArrayList<GCalendarDate>();
		try {
			SimpleDateFormat DATE_FORMAT = new SimpleDateFormat("yyyyMMdd");
			Calendar startDate = Calendar.getInstance();
			startDate.setTime(DATE_FORMAT.parse(String.valueOf(this.start_date)));
			Calendar endDate = Calendar.getInstance();
			endDate.setTime(DATE_FORMAT.parse(String.valueOf(this.end_date)));
			Calendar c = startDate; // no need to clone because not re-using startDate later
			c.add(Calendar.DAY_OF_MONTH, -1); // starting yesterday because increment done at the beginning of the loop
			int date;
			while (c.before(endDate)) {
				c.add(Calendar.DAY_OF_MONTH, +1);
				try {
					date = Integer.valueOf(DATE_FORMAT.format(c.getTime()));
					switch (c.get(Calendar.DAY_OF_WEEK)) {
					case Calendar.MONDAY:
						if (this.monday) {
							this.allDates.add(new GCalendarDate(this.service_id, date, GCalendarDatesExceptionType.SERVICE_ADDED));
							continue; // service today
						}
						break;
					case Calendar.TUESDAY:
						if (this.tuesday) {
							this.allDates.add(new GCalendarDate(this.service_id, date, GCalendarDatesExceptionType.SERVICE_ADDED));
							continue; // service today
						}
						break;
					case Calendar.WEDNESDAY:
						if (this.wednesday) {
							this.allDates.add(new GCalendarDate(this.service_id, date, GCalendarDatesExceptionType.SERVICE_ADDED));
							continue; // service today
						}
						break;
					case Calendar.THURSDAY:
						if (this.thursday) {
							this.allDates.add(new GCalendarDate(this.service_id, date, GCalendarDatesExceptionType.SERVICE_ADDED));
							continue; // service today
						}
						break;
					case Calendar.FRIDAY:
						if (this.friday) {
							this.allDates.add(new GCalendarDate(this.service_id, date, GCalendarDatesExceptionType.SERVICE_ADDED));
							continue; // service today
						}
						break;
					case Calendar.SATURDAY:
						if (this.saturday) {
							this.allDates.add(new GCalendarDate(this.service_id, date, GCalendarDatesExceptionType.SERVICE_ADDED));
							continue; // service today
						}
						break;
					case Calendar.SUNDAY:
						if (this.sunday) {
							this.allDates.add(new GCalendarDate(this.service_id, date, GCalendarDatesExceptionType.SERVICE_ADDED));
							continue; // service today
						}
						break;
					}
				} catch (Exception e) {
					System.out.printf("\nError while parsing date '%s'!\n", c);
					e.printStackTrace();
					System.exit(-1);
				}
			}
		} catch (Exception e) {
			System.out.printf("\nError while parsing dates between '%s' and '%s'!\n", this.start_date, this.end_date);
			e.printStackTrace();
			System.exit(-1);
		}
	}
}

File: src/main/java/org/mtransit/parser/gtfs/data/GCalendar.kt
package org.mtransit.parser.gtfs.data

import androidx.annotation.Discouraged
import org.mtransit.parser.MTLog
import org.mtransit.parser.db.SQLUtils.escapeId
import java.util.Calendar

// https://developers.google.com/transit/gtfs/reference#calendar_fields
// https://gtfs.org/reference/static/#calendartxt
data class GCalendar(
    val serviceIdInt: Int,
    private val monday: Boolean,
    private val tuesday: Boolean,
    private val wednesday: Boolean,
    private val thursday: Boolean,
    private val friday: Boolean,
    private val saturday: Boolean,
    private val sunday: Boolean,
    val startDate: Int, // YYYYMMDD
    val endDate: Int, // YYYYMMDD
) {
    constructor(
        serviceId: String,
        monday: Boolean,
        tuesday: Boolean,
        wednesday: Boolean,
        thursday: Boolean,
        friday: Boolean,
        saturday: Boolean,
        sunday: Boolean,
        startDate: Int,
        endDate: Int,
    ) : this(
        GIDs.getInt(serviceId),
        monday,
        tuesday,
        wednesday,
        thursday,
        friday,
        saturday,
        sunday,
        startDate,
        endDate
    )

    constructor(
        serviceId: String,
        monday: Int,
        tuesday: Int,
        wednesday: Int,
        thursday: Int,
        friday: Int,
        saturday: Int,
        sunday: Int,
        startDate: Int,
        endDate: Int,
    ) : this(
        GIDs.getInt(serviceId),
        monday == 1,
        tuesday == 1,
        wednesday == 1,
        thursday == 1,
        friday == 1,
        saturday == 1,
        sunday == 1,
        startDate,
        endDate
    )

    @Suppress("unused")
    @get:Discouraged(message = "Not memory efficient")
    val serviceId: String get() = _serviceId

    private val _serviceId: String
        get() = GIDs.getString(serviceIdInt)

    val escapedServiceId: String
        get() = _serviceId.escapeId()

    val escapedServiceIdInt: Int
        get() = escapedServiceId.toGIDInt()

    val dates: List<GCalendarDate> by lazy {
        initAllDates(
            serviceIdInt,
            monday,
            tuesday,
            wednesday,
            thursday,
            friday,
            saturday,
            sunday,
            startDate,
            endDate
        )
    }

    @JvmOverloads
    fun flattenToCalendarDates(
        exceptionType: GCalendarDatesExceptionType = GCalendarDatesExceptionType.SERVICE_DEFAULT,
    ) =
        initAllDates(
            serviceIdInt,
            monday,
            tuesday,
            wednesday,
            thursday,
            friday,
            saturday,
            sunday,
            startDate,
            endDate,
            exceptionType
        )

    fun hasDays(): Boolean {
        return monday || tuesday || wednesday || thursday || friday || saturday || sunday
    }

    fun isServiceIdInts(serviceIdInts: Collection<Int?>): Boolean {
        return serviceIdInts.contains(serviceIdInt)
    }

    @Suppress("unused")
    fun startsBefore(date: Int): Boolean {
        return startDate < date
    }

    @Suppress("MemberVisibilityCanBePrivate")
    fun startsBetween(startDate: Int, endDate: Int): Boolean {
        return this.startDate in startDate..endDate
    }

    @Suppress("unused")
    fun isOverlapping(startDate: Int?, endDate: Int?): Boolean {
        startDate ?: return false
        endDate ?: return false
        return startsBetween(startDate, endDate) || endsBetween(startDate, endDate)
    }

    @Suppress("unused")
    fun isInside(startDate: Int?, endDate: Int?): Boolean {
        startDate ?: return false
        endDate ?: return false
        return startsBetween(startDate, endDate) && endsBetween(startDate, endDate)
    }

    @Suppress("MemberVisibilityCanBePrivate")
    fun endsBetween(startDate: Int, endDate: Int): Boolean {
        return this.endDate in startDate..endDate
    }

    @Suppress("unused")
    fun endsAfter(date: Int): Boolean {
        return endDate > date
    }

    @Suppress("unused")
    fun containsDate(date: Int?): Boolean {
        date ?: return false
        return date in startDate..endDate
    }

    @Suppress("unused")
    fun toStringPlus(): String {
        return toString() +
                "+(serviceIdInt:$_serviceId)"
    }

    @Suppress("unused")
    fun toStringShort() = buildString {
        append("Calendar:{'")
        append(_serviceId)
        append("': ")
        append(startDate)
        append("-")
        append(endDate)
        append(" [")
        append(if (monday) "M" else "_")
        append(if (tuesday) "T" else "_")
        append(if (wednesday) "W" else "_")
        append(if (thursday) "T" else "_")
        append(if (friday) "F" else "_")
        append(if (saturday) "S" else "_")
        append(if (sunday) "S" else "_")
        append("]")
        append("}")
    }

    fun isRunningOnCalendarDayOfWeek(calendarDayOfWeek: Int) =
        when (calendarDayOfWeek) {
            Calendar.MONDAY -> monday
            Calendar.TUESDAY -> tuesday
            Calendar.WEDNESDAY -> wednesday
            Calendar.THURSDAY -> thursday
            Calendar.FRIDAY -> friday
            Calendar.SATURDAY -> saturday
            Calendar.SUNDAY -> sunday
            else -> {
                MTLog.log("Unexpected day of week '$calendarDayOfWeek'!")
                false
            }
        }

    companion object {
        const val FILENAME = "calendar.txt"

        internal const val SERVICE_ID = "service_id"
        private const val MONDAY = "monday"
        private const val TUESDAY = "tuesday"
        private const val WEDNESDAY = "wednesday"
        private const val THURSDAY = "thursday"
        private const val FRIDAY = "friday"
        private const val SATURDAY = "saturday"
        private const val SUNDAY = "sunday"
        private const val START_DATE = "start_date"
        private const val END_DATE = "end_date"

        private const val DAY_TRUE = "1"

        @JvmStatic
        fun fromLine(line: Map<String, String>) = GCalendar(
            line[SERVICE_ID] ?: throw MTLog.Fatal("Invalid GCalendar from $line!"),
            DAY_TRUE == line[MONDAY],
            DAY_TRUE == line[TUESDAY],
            DAY_TRUE == line[WEDNESDAY],
            DAY_TRUE == line[THURSDAY],
            DAY_TRUE == line[FRIDAY],
            DAY_TRUE == line[SATURDAY],
            DAY_TRUE == line[SUNDAY],
            line[START_DATE]?.toInt() ?: throw MTLog.Fatal("Invalid GCalendar from $line!"),
            line[END_DATE]?.toInt() ?: throw MTLog.Fatal("Invalid GCalendar from $line!"),
        )

        fun isRunningOnDay(calendar: GCalendar, dayString: String): Boolean =
            calendar.isRunningOnCalendarDayOfWeek(
                Calendar.getInstance()
                    .apply { time = GFieldTypes.makeDateFormat().parse(dayString) }
                    .get(Calendar.DAY_OF_WEEK))

        @JvmStatic
        @JvmOverloads
        fun flattenToCalendarDates(
            calendars: Collection<GCalendar>,
            exceptionType: GCalendarDatesExceptionType = GCalendarDatesExceptionType.SERVICE_DEFAULT,
        ) = calendars.flatMap { it.flattenToCalendarDates(exceptionType) }

        private fun initAllDates(
            serviceId: Int,
            monday: Boolean,
            tuesday: Boolean,
            wednesday: Boolean,
            thursday: Boolean,
            friday: Boolean,
            saturday: Boolean,
            sunday: Boolean,
            startDate: Int,
            endDate: Int,
            exceptionType: GCalendarDatesExceptionType = GCalendarDatesExceptionType.SERVICE_DEFAULT,
        ) = buildList {
            try {
                val dateFormat = GFieldTypes.makeDateFormat()
                val startDateCal = Calendar.getInstance().apply {
                    time = dateFormat.parse(startDate.toString())
                }
                val endDateCal = Calendar.getInstance().apply {
                    time = dateFormat.parse(endDate.toString())
                }
                val c = startDateCal // no need to clone because not re-using startDate later
                c.add(Calendar.DAY_OF_MONTH, -1) // starting yesterday because increment done at the beginning of the loop
                while (c.before(endDateCal)) {
                    c.add(Calendar.DAY_OF_MONTH, +1) // beginning of the loop
                    try {
                        val date = Integer.valueOf(dateFormat.format(c.time))
                        when (c[Calendar.DAY_OF_WEEK]) {
                            Calendar.MONDAY -> if (monday) add(GCalendarDate(serviceId, date, exceptionType))
                            Calendar.TUESDAY -> if (tuesday) add(GCalendarDate(serviceId, date, exceptionType))
                            Calendar.WEDNESDAY -> if (wednesday) add(GCalendarDate(serviceId, date, exceptionType))
                            Calendar.THURSDAY -> if (thursday) add(GCalendarDate(serviceId, date, exceptionType))
                            Calendar.FRIDAY -> if (friday) add(GCalendarDate(serviceId, date, exceptionType))
                            Calendar.SATURDAY -> if (saturday) add(GCalendarDate(serviceId, date, exceptionType))
                            Calendar.SUNDAY -> if (sunday) add(GCalendarDate(serviceId, date, exceptionType))
                        }
                    } catch (e: Exception) {
                        throw MTLog.Fatal(e, "Error while parsing date '$c'!")
                    }
                }
            } catch (e: Exception) {
                throw MTLog.Fatal(e, "Error while parsing dates between '$startDate' and '$endDate'!")
            }
        }
    }
}